src/Shared/Buffers.MemoryPool/PinnedBlockMemoryPool.cs

```diff
@@ -82,12 +134,94 @@ internal void Return(MemoryPoolBlock block)
             block.IsLeased = false;
 #endif
 
+        Interlocked.Increment(ref _returnCount);
+
         if (!_isDisposed)
         {
+            _metrics?.UpdateCurrentMemory(block.Memory.Length);
+            Interlocked.Add(ref _currentMemory, block.Memory.Length);
+
             _blocks.Enqueue(block);
         }
     }
 
+    public bool TryScheduleEviction(DateTimeOffset now)
+    {
+        if (now >= _nextEviction)
+        {
+            _nextEviction = now.Add(DefaultEvictionDelay);
+            ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+            return true;
+        }
+
+        return false;
+    }
+
+    void IThreadPoolWorkItem.Execute()
+    {
+        try
+        {
+            PerformEviction();
+        }
+        catch (Exception ex)
+        {
+            // Log the exception, but don't let it crash the thread pool
+            _logger?.LogError(ex, "Error while performing eviction in PinnedBlockMemoryPool.");
```

Comment:
```suggestion
            _logger?.LogCritical(ex, "Error while performing eviction in PinnedBlockMemoryPool.");
```

I think a critical log is appropriate considering something must have gone horribly wrong to get here.